Xamarin/AndroidClient/MainActivity.cs

```diff
@@ -18,74 +28,160 @@ protected override void OnCreate(Bundle savedInstanceState)
+            //alertDialogBuilder.SetPositiveButton("Ok", (_, __) =>
+            //{
+            //    userName = nameInput.Text;
+            //});
+
+            //alertDialogBuilder.SetNegativeButton("Cancel", (_, __) => { });
+
+            //alertDialogBuilder.Show();
+        }
+
+        private async void ConnectButton_Click(object sender, EventArgs e)
+        {
+            if (_state == ViewState.Connected)
             {
-                if (!connected)
+                UpdateState(ViewState.Disconneting);
+
+                try
+                {
+                    await _connection.StopAsync();
+                }
+                catch (Exception ex)
                 {
-                    Toast.MakeText(this, "Must be Connected to Send!", ToastLength.Short).Show();
+                    AppendMessage($"An error occurred while disconnecting: {ex}");
+                    UpdateState(ViewState.Connected);
                     return;
                 }
 
+                UpdateState(ViewState.Disconnected);
+            }
+            else
+            {
+                UpdateState(ViewState.Connecting);
+
                 try
                 {
-                    await hubConnection.InvokeAsync("Send", userName, messageText.Text);
-                    messageText.Text = "";
+                    _connection = new HubConnectionBuilder()
+                        .WithUrl(_serverUrlText.Text)
+                        .ConfigureLogging(logging => { logging.AddDebug(); })
+                        .Build();
+
+                    _connection.On<string>("Send", (message) =>
+                    {
+                        AppendMessage(message);
+                    });
+
+                    await _connection.StartAsync();
                 }
                 catch (Exception ex)
                 {
-                    var exception = ex;
+                    AppendMessage($"An error occurred while connecting: {ex}");
+                    UpdateState(ViewState.Disconnected);
                 }
-            };
+
+                UpdateState(ViewState.Connected);
+            }
+        }
+
+        private async void SendButton_Click(object sender, EventArgs e)
+        {
+            if (_state != ViewState.Connected)
+            {
+                Toast.MakeText(this, "Must be Connected to Send!", ToastLength.Short).Show();
+                return;
+            }
+
+            try
+            {
+                await _connection.InvokeAsync("Send", _messageText.Text);
+                _messageText.Text = "";
+            }
+            catch (Exception ex)
+            {
+                AppendMessage($"An error occurred while sending: {ex}");
+            }
+        }
+
+        private void UpdateState(ViewState state)
+        {
+            if (_state == state)
+            {
+                return;
+            }
+
+            switch (state)
+            {
+                case ViewState.Connected:
```

Comment:
You could simplify this a little. Like if it's not in the connected state then the `sendButton` and `messageTest` are never enabled. That makes the switch considerably smaller

Reply:
Yeah, but in this case I think I'd prefer it's clear about the state of the button in each view state.